### ISSUE TITLE
SearchPanes support

### ISSUE THREAD

**Issue by brobey8** [yajra/laravel-datatables#2463]
### Summary of problem or feature request

When including the SearchPane datatables plugin (https://datatables.net/blog/2020-01-13) it looks like there is no data rendered.. what is needed to implement this functionality inside Yajra?

**Comment by schonhoff** [comment]
Do you use server-side processing?
Currently the server-side processing of this project doesn't support the needed data shown in:
https://datatables.net/extensions/searchpanes/serverside

Maybe you can find a way by yourself to add the needed data to your datatables response.

**Comment by brobey8** [comment]
@schonhoff Yep I am using server side processing (also using the "service implementation" which is really nice). 

> Maybe you can find a way by yourself to add the needed data to your datatables response.

I believe that I will be doing this as my client wants the extra functionality offered by the SearchPanes, I will try doing it myself using the DataTables API until the feature is supported.

**Comment by yajra** [comment]
Would definitely support this feature soon. Maybe by next month when hacktoberfest begins ^_^

Accepting PR though incase someone wants to implement it. Thanks!

**Comment by brobey8** [comment]
@yajra That's good to hear mate, if I end up extending the package to achieve this I will submit a PR :)

**Comment by pacmac0** [comment]
Tldr;
Is there a posibility at the moment to get the searchpanes with server side processed data from an Ajax call running?
So far I get just 'No data available in table'.
From this threat it seems like a solution was created, how can I get this?
Sorry I am fairly new and inexperienced, would appreciate help a lot!!!

Hello everyone,
I want to use the searchpanes functionality for my server side rendered data, but I get 'No data available in table' in the shown panes.
I am fairly new to datatables, but as I understood from this issue, it doesn't support server side rendering with searchpanes yet?
I also found this example(https://editor.datatables.net/examples/extensions/searchPanes.html), where it seems that there is a way, but I don't really understand what is happening.
My question now, how can I use the build in searchpanes with server side processed data from an Ajax call?
Anyone found a solution for the time being?
Thanks for your help guys, I appreciate it a lot!!

**Comment by schonhoff** [comment]
Tldr;
It is coming, but don't know when.
Documentation for server side loading on SearchPanes: https://datatables.net/extensions/searchpanes/serverside

Hello,

SearchPanes will be integrated in the future. Yajra is currently working on it in the pull request:
https://github.com/yajra/laravel-datatables/pull/2475
I don't know how far the pull request currently is. But if it is finished it will include SearchPanes.

In the example you linked you can watch the Ajax load after clicking in the SearchPanes. There you should notice a  "searchPanes" object. This needs to be done server side to load the SearchPanes. Here is the documentation from datatables SearchPanes extension with server side loading:
https://datatables.net/extensions/searchpanes/serverside

Hope I could help you a little bit.

**Comment by pacmac0** [comment]
Thank you,
So just that I get it right, currently I am returning the column row definitions as Json with the  Datatables::of($my_eloquent_model) function. To use the searchpanes I have to add this searchpanes object as discribed in (https://datatables.net/extensions/searchpanes/serverside) right? To create this serverside, I guess to do so I have to use the 'DataTables\Editor' package as shown in (https://editor.datatables.net/examples/extensions/searchPanes.html). But the example code looks so different, I don't understand how they relate. Do I have to create the searchpane object indipendent and add it to the json response off the call? 
And do i have to handel the reload somehow when the user selects a searchpane?

Thanks a lot for your help!

**Comment by schonhoff** [comment]
Hi,

`Do I have to create the searchpane object indipendent and add it to the json response off the call?`
Yes, that would be the case if you wanna implement it already instead waiting for the new release from Yajra.

`And do i have to handel the reload somehow when the user selects a searchpane?`
Yes, you have to handel the reload by yourself. You can do it already with the filter option and the keyword needs to be changed to the searchpane request input. This is not visibile in the editor example you mentioned. The example is btw broken on the datatables site. I would recommend looking at the request by your own. So you can look up what request is given back to your server side script.

In my opinion waiting for it would be the best option. If you want to have it already implemented just use client side Datatables. That works already with this package. I used it quite a lot in another project like this because it is not implemented server side at the moment. If it is implemented server side I will switch the functionality back to server side.

**Comment by pacmac0** [comment]
Ok you are right, that seems like a lot of extra effort.
As I said I am really a newby, how long do you think it will take until the feature is available in a new version? (I know it will just be an estimate, but at least an educated one :D).

Cheers pacmac

**Comment by schonhoff** [comment]
Yajra is already on it and he marked it with the Hacktoberfest banner. So i would guess he is working on it. But maybe there are problems that I don't know at the moment. In my opinion such a big change would take more time. Like ~1 month more? But I can't really tell how far Yajra currently is.

**Comment by pacmac0** [comment]
Ok 
Thank you!

**Comment by yajra** [comment]
To keep everyone here updated, the only thing missing on https://github.com/yajra/laravel-datatables/pull/2475 is the ability to apply a custom search handler. I already have an implementation plan but can't find time to work on it yet as several work projects came in simultaneously. Original target to complete this feature is within hacktoberfest. So hopefully, I can get back on track and complete this the soonest >.< 🍻

**Comment by schonhoff** [comment]
Thanks for the update. Take your time 😃 
Maybe I can take some time and take a look into the PR next week and help. But I can't promise anything. :-(

**Comment by yajra** [comment]
SearchPanes support released on v9.13.0 and v4.34.0 for html builder. 🚀 🍺 

Note: **NO** docs yet. See PR #2470 and https://github.com/yajra/laravel-datatables-html/pull/137 for details. Accepting PR on [docs repo](https://github.com/yajra/laravel-datatables-docs) 😄

**Comment by vanthao03596** [comment]
@yajra I tried SearchPances but it always show No SearchPances. Here is my code

```php
<?php

namespace App\DataTables;

use App\Models\User;
use Yajra\DataTables\Html\Button;
use Yajra\DataTables\Html\Column;
use Yajra\DataTables\Html\Editor\Editor;
use Yajra\DataTables\Html\Editor\Fields;
use Yajra\DataTables\Html\Editor\Fields\Field;
use Yajra\DataTables\Services\DataTable;

class UsersDataTable extends DataTable
{
    /**
     * Build DataTable class.
     *
     * @param mixed $query Results from query() method.
     * @return \Yajra\DataTables\DataTableAbstract
     */
    public function dataTable($query)
    {
        $users = User::query()->select('id as value', 'name as label')->get();
        return datatables()
            ->eloquent($query)
            ->setRowId('id')
            ->searchPane('user_id', $users)
            ->addColumn('action', 'users.action');
    }

    /**
     * Get query source of dataTable.
     *
     * @param \App\Models\User $model
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function query(User $model)
    {
        return $model->newQuery();
    }

    /**
     * Optional method if you want to use html builder.
     *
     * @return \Yajra\DataTables\Html\Builder
     */
    public function html()
    {
        return $this->builder()
                    ->setTableId('users-table')
                    ->columns($this->getColumns())
                    ->minifiedAjax()
                    ->dom('PBfrtip')
                    ->orderBy(1)
                    ->selectStyleMulti()
                    ->selectSelector('td:not(:last-child)')
                    ->searchPanes(true)
                    ->buttons(
                        Button::make('add'), // appends /create on current URL.
                        Button::make('create')->editor('editor'),
                        Button::make('edit')->editor('editor'),
                        Button::make('remove')->editor('editor'),
                        Button::make('restore'),
                        Button::make('duplicate'),
                        Button::make('duplicateSingle'),
                        Button::make('collection')
                            ->buttons([
                                Button::make('forceDelete'), // Force delete record(/s).
                                Button::make('forceDeleteSingle'), // Force delete a single record.
                            ])
                            ->text('Action'),
                            Button::make('toggleScope')
                                 ->text('<i class="fa fa-square"></i> Only Deleted')
                                ->scope('onlyDeleted'),
                        Button::make('export'),
                        Button::make('print'),
                        Button::make('reset'),
                        Button::make('reload'),
                    )
                    ->editor(
                        Editor::make()
                            ->fields([
                                Fields\Text::make('name'),
                                Fields\Text::make('email'),
                                Fields\Password::make('password'),
                                Fields\Boolean::make('is_active'),
                            ])
                    );
    }

    /**
     * Get columns.
     *
     * @return array
     */
    protected function getColumns()
    {
        return [
            Column::checkbox(''),
            Column::make('id')->title(__('Index'))->data('DT_RowIndex')->searchable(false),
            Column::make('email'),
            Column::make('name'),
            Column::make('created_at'),
            Column::make('updated_at'),
            Column::make('deleted_at')->renderJs('boolean()'),
            Column::computed('action')
            ->exportable(false)
            ->printable(false)
            ->width(60)
            ->addClass('text-center'),
        ];
    }

    /**
     * Get filename for export.
     *
     * @return string
     */
    protected function filename()
    {
        return 'Users_' . date('YmdHis');
    }
}

```

**Comment by yajra** [comment]
@vanthao03596  wow that's full featured code you got there. It's very rare for me to see snippets like this in the community as it is not fully documented. ^_^

Anyways, try this approach of using the searchPanes buttons. 

```php
    public function html()
    {
        return $this->builder()
                    ->setTableId('users-table')
                    ->columns($this->getColumns())
                    ->minifiedAjax()
                    ->dom('Bfrtip') // remove P
                    ->orderBy(1)
                    ->selectStyleMulti()
                    ->selectSelector('td:not(:last-child)')
                    // ->searchPanes(true) this is optional on server side
                    ->buttons(
                        Button::make('searchPanes'),
```

**Comment by yajra** [comment]
Also you might want to hide the count on search panes.

```php
->searchPanes(['hideCount' => true])
```

**Comment by vanthao03596** [comment]
@yajra I tried and it still show No SearchPanes 
![image](https://user-images.githubusercontent.com/34786441/99361152-3f9f6f80-28e4-11eb-8ba6-dde23cf9bd51.png)

**Comment by yajra** [comment]
Your searchPanes returns `created_by` key. It should be `user_id` or any columns that matches your column definitions name.

**Comment by vanthao03596** [comment]
@yajra my column name is created_by. I created repositoy at [https://github.com/vanthao03596/laravel-datatable-searchpanes](here). Can you take a look at my code sir ?

**Comment by yajra** [comment]
This is a weird behavior of searchPanes but I was able to make your code works by adding a column definition:

```php
->addColumnDef([
  'searchPanes' => ['show' => true],
  'targets' => [7]
])
```

On my current project, I don't need to include a column definition at all and it just works 😄 

P.S. I suggest you delete your repo afterwards because the editor scripts were exposed and it should not be publicly available.

**Comment by vanthao03596** [comment]
Thanks you so much. Now i can use SearchPanes.

**Comment by vanthao03596** [comment]
Hi @yajra . Do you have any idea to use SearchPanes Integration for Editor with CascadePanes in serverside.

**Comment by yajra** [comment]
Afaik, cascade panes is not yet supported on serverSide and haven't tried it yet. 

Anyways, any demo link where I can see what you want to build? Will try it when I got the chance.

**Comment by vanthao03596** [comment]
@yajra  See this demo:
https://editor.datatables.net/examples/extensions/searchPanesViewTotal.html

**Comment by SpenzeR** [comment]
I found a solution that solved my server-side searchPane issue.

```
return Datatables::of(.....
.....
->with([
    'searchPanes' => [
        'options' => [
            'name' => [
                [
                    'label' => 'name',
                    'total' => User::where('name', 'Ola Nordmann')->count(),
                    'value' => 'John Doe',
                ]
            ]
        ]
    ]
])
.....
.....
```

But using the above code would still give me a server-side error (undefined index "name").

To solve this error I had to modify the following file:

> src/vendor/yajra/laravel-datatables-oracle/src/QueryDataTable.php

I changed line 122 (inside searchPanesSearch function)

From:
`if ($this->searchPanes[$column] && $callback = $this->searchPanes[$column]['builder']) {`

To:
`if (isset($this->searchPanes[$column]) && $callback = $this->searchPanes[$column]['builder']) {`

Hope this is to any help.

**Comment by aweiand** [comment]
Any support now? I didin`t find any documentation about this yet....

**Comment by psmithne** [comment]
Hi @yajra  Great package.  SearchPanes is working well with my server side implementation.  Like others, I would also love to see the CascadePanes option supported.  While Total Counts are working well, users expect their selections to be cumulative/cascade/self filter.

**Comment by MuhammadTanvir** [comment]
Hi @yajra, I tried to use the SearchPanes on the serverSide datatable, but I doesn't work, the SearchPanes options are empty.

Here is my codes-

`public function getSalesReturnsData()
{
    
    $sales= Sale::query()
            ->select(DB::raw("Bestelldatum, kLieferscheinPos ,cBestellNr,
                Verkaufsplattform, SKU, GTIN, HAN, Hersteller, Lieferant, cName, Menge, VKPreis, UVP, Farbe,
                Groesse, Warengruppe, Transaktion, Anmerkung, Lieferland, Versandart, Zahlungsart, VersandkostenBrutto, GutscheinBestellungBrutto"))
                ->where('Bestelldatum', '>', now()->subDays(30)->endOfDay());

    $returns= ProductReturn::query()
            ->select(DB::raw("Retourendatum, kRMRetoure,cBestellNr, Verkaufsplattform, SKU, GTIN, HAN, Hersteller, Lieferant, cName, Menge, VKPreis, UVP, Farbe,
                Groesse, Warengruppe, Transaktion, Anmerkung, Lieferland, Versandart, Zahlungsart, VersandkostenBrutto, GutscheinBestellungBrutto"));

    $results = $sales->unionAll($returns);


    return DataTables::of($results)
        ->addColumn('Bestelldatum', function($row){
                $bestelldatum = date("d.m.Y", strtotime($row->Bestelldatum));
                return $bestelldatum;
        })->make(true);
}`

`<script>
    $(document).ready(function() {

        load_data();

        function load_data(){
            $('#datatable').DataTable({
                processing: true,
                serverSide: false,
                ajax: {
                    url: "{{ route('getSalesReturnsData') }}",
                },
                "columns": [
                    {
                        "data": "Hersteller"
                    },
                ],
                dom: 'PBfrtip',
                stateSave: true,
                searchPanes: {
                    initCollapsed: true,
                    cascadePanes: true,
                    dtOpts: {
                        paging: 'flase',
                        pagingType: 'simple',
                        select: {
                            style: 'multi'
                        }
                    },
                    columns: [3, 5, 8, 6, 7, 9, 11, 16]
                }
            });
        }

    });
</script>`
